velox/exec/PrefixSort.h

```diff
@@ -0,0 +1,150 @@
+  // position when extract columns.
+  std::vector<int32_t> prefixOffsets_;
+};
+
+// Put inline methods in the header file to avoid link error. Although
+// PrefixSortLayout is used only in-side PrefixSort, it`s a data member, it
+// should be full defined when compile PrefixSort.
+FOLLY_ALWAYS_INLINE uint32_t PrefixSortLayout::normalizedKeySize() const {
+  return normalizedKeySize_;
+}
+
+FOLLY_ALWAYS_INLINE uint64_t PrefixSortLayout::entrySize() const {
+  return entrySize_;
+}
+
+FOLLY_ALWAYS_INLINE int32_t PrefixSortLayout::numNormalizedKeys() const {
+  return numNormalizedKeys_;
+}
+
+FOLLY_ALWAYS_INLINE int32_t PrefixSortLayout::numSortKeys() const {
+  return numSortKeys_;
+}
+
+FOLLY_ALWAYS_INLINE CompareFlags
+PrefixSortLayout::compareFlag(const uint32_t index) const {
+  return keyCompareFlags_[index];
+}
+
+FOLLY_ALWAYS_INLINE int32_t
+PrefixSortLayout::prefixOffset(const uint32_t index) const {
+  return prefixOffsets_[index];
+}
+
+FOLLY_ALWAYS_INLINE bool PrefixSortLayout::hasNonNormalizedKey() const {
+  return hasNonNormalizedKey_;
+}
+
+class PrefixSort {
+ public:
+  PrefixSort(
+      memory::MemoryPool* pool,
+      RowContainer* rowContainer,
+      const std::vector<CompareFlags>& keyCompareFlags,
+      const PrefixSortConfig& config);
+
+  /// Sort the rows of a RowContainer with steps:
+  /// 1. Allocates a contiguous memory block to store the normalized keys.
+  /// 2. Extract sort keys of rows from RowContainer, encode them to
+  /// normalized-form which can be normalized, the normalized-form data with
+  /// the row address is called 'Prefix'.
+  /// 3. Sort the prefixes data we got in step 2.
+  /// For keys can normalize(All fixed width types), we use memcmp to compare
+  /// the normalized binary string.
+  /// For keys can not normalize, we use RowContainer`s compare method to
+  /// compare value.
+  /// For keys can part-normalize(Varchar, Row etc.), we will store the
+  /// normalized part and points to raw data in prefix, and custom the points
+  /// compare. The compare strategy will be defined in PrefixSortLayout as
+  /// follow-up, we trait this part as non-normalize until we implement all
+  /// fixed width types.
+  /// For complex types, ROW that could represent as Scalar types will be
+  /// supported.
+  /// 4. Extract the row address from sorted prefixes into input rows vector.
+  ///
+  /// @param rows The result of RowContainer::listRows(), assuming that the
+  /// caller (SortBuffer etc.) has already got the result.
+  void sort(std::vector<char*>& rows);
+
+ private:
+  int compareAllNormalizedKeys(char* left, char* right);
+
+  int comparePartNormalizedKeys(char* left, char* right);
+
+  void extractRowToPrefix(char* row, char* prefix);
+
+  FOLLY_ALWAYS_INLINE char* getAddressFromPrefix(char* prefix) {
+    return *reinterpret_cast<char**>(prefix + sortLayout_.normalizedKeySize());
+  }
+
+  memory::MemoryPool* pool_;
+  PrefixSortLayout sortLayout_;
+  RowContainer* rowContainer_;
```

Comment:
What methods of the RowContainer do we need? Can we remove dependency on RowContainer by taking a function to compare rows instead?

Reply:
Also need RowColumn & RowContainer::isNullAt